doc/lexer_tutorial.md

```diff
@@ -0,0 +1,192 @@
+# Writing a custom lexer
+
+Let's say we want to parse the Whitespace language, so we've put together a grammar like the following:
+
+```rust
+pub Program = <Statement*>;
+
+Statement: ast::Stmt = {
+    " " <StackOp>,
+    "\t" " " <MathOp>,
+    "\t" "\t" <HeapOp>,
+    "\n" <FlowCtrl>,
+    "\t" "\n" <Io>,
+};
+
+StackOp: ast::Stmt = {
+    " " <Number> => ast::Stmt::Push(<>),
+    "\n" " " => ast::Stmt::Dup,
+    "\n" "\t" => ast::Stmt::Swap,
+    "\n" "\n" => ast::Stmt::Discard,
+};
+
+MathOp: ast::Stmt = {
+    " " " " => ast::Stmt::Add,
+    " " "\t" => ast::Stmt::Sub,
+    " " "\n" => ast::Stmt::Mul,
+    "\t" " " => ast::Stmt::Div,
+    "\t" "\t" => ast::Stmt::Mod,
+};
+
+// Remainder omitted
+```
+
+Naturally, it doesn't work. By default, LALRPOP generates a tokenizer that skips all whitespace -- including newlines. What we *want* is to capture whitespace characters and ignore the rest as comments, and LALRPOP does the opposite of that.
+
+At the moment, LALRPOP doesn't allow you to configure the default tokenizer. In the future it will become quite flexible, but for now we have to write our own.
+
+Let's start by defining the stream format. The parser will accept an iterator where each item in the stream has the following structure:
+
+```rust
+pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;
+```
+
+`Loc` is typically just a `usize`, representing a byte offset into the input string. Each token is accompanied by two of them, marking the start and end positions where it was found. `Error` can be pretty much anything you choose. And of course `Tok` is the meat of the stream, defining what possible values the tokens themselves can have. Following the conventions of Rust iterators, we'll signal a valid token with `Some(Ok(...))`, an error with `Some(Err(...))`, and EOF with `None`.
+
+(Note that the term "tokenizer" normally refers to a piece of code that simply splits up the stream, whereas a "lexer" also tags each token with its lexical category. What we're writing is the latter.)
+
+Whitespace is a simple language from a lexical standpoint, with only three valid tokens:
+
+```rust
+pub enum Tok {
+    Space,
+    Tab,
+    Linefeed,
+}
+```
+
+Everything else is a comment. There are no invalid lexes, so we'll define our own error type, a void enum:
+
+```rust
+pub enum LexicalError {
```

Comment:
Nit: I *think* the error can be omitted if the lexer is infallible -- but then I like including it, since most lexers *can* fail. And I might have taken away the ability to omit it, I forget. :)